File: Solutions/BinarySearch.cpp
// Leetcode #704: https://leetcode.com/problems/binary-search/

// Solution 1 | O(logn) time, O(1) space
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // Search space is the whole vector
        int left = 0, right = nums.size() - 1;
        
        // While there are elements in the search space
        while (left <= right) {
            // Calculate the middle index
            int mid = left + (right - left) / 2;

            // If it's too large, discard the right half
            if (nums[mid] > target) right = mid - 1;
            
            // Else if it's too small, discard the left half
            else if (nums[mid] < target) left = mid + 1;
            
            // Else we found the target, return its index
            else return mid;
        }
        
        return -1; // If we reach here, target was not found
    }
};

File: Solutions/BinarySearch.py
# Leetcode #704: https://leetcode.com/problems/binary-search/description/

# Solution 1 | O(logn) time, O(1) space
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # Left and right pointers denote the search range
        left, right = 0, len(nums) - 1

        # While the search range is non-empty, find the middle index and check if its the target
        # If nums[mid] is too small, discard the left half. If its too big, discard the right half
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                left = mid + 1
            else:
                right = mid - 1

        # Target not found, return -1
        return -1
